File: Problems 10-19/problem10.py
def sieve_of_eratosthenes(n = 10001):
	n = n + 1
	sieve = range(n)
	sieve[:2] = [0, 0]

	for i in xrange(2, int(n**0.5) + 1):
		if sieve[i]:
			for j in xrange(i**2, n, i):
				sieve[j] = 0

	#filter out the values that aren't set to zero
	return [p for p in sieve if p]

def sum_of_primes_below(n = 2000000):
	sum_of_primes = sum(sieve_of_eratosthenes(n))
	return sum_of_primes

if __name__ == "__main__":
	target = 2000000
	print "The sum of all primes less than %d is %d" % (target, sum_of_primes_below(target))

File: Problems 10-19/problem10.c
#include <stdio.h>
#include <stdlib.h> 

unsigned long sum_sieve_of_eratosthenes(unsigned long n);

int main()
{
	unsigned long n = 2000000;

	printf("The sum of all primes less than %lu is: %lu\n", 
		n, sum_sieve_of_eratosthenes(n));

	return 0;
}

/**
 * Slight modification to the sieve to sum all primes rather
 * than returning the first value which meets our upper limit
 */
unsigned long sum_sieve_of_eratosthenes(unsigned long n)
{
	char *sieve;
	unsigned long i, result = 0, upper_limit = 1000000;

	sieve = calloc(n, sizeof *sieve); //Initialises to zero

	for (i = 2; i < upper_limit; i++)
	{
		if (!sieve[i])
		{
			unsigned long j;
			result += i;

			for (j = i*2; j < upper_limit; j += i)
			{
				sieve[j] = 1;
			}
		}
	}

	free(sieve); //Clear up our memory

	return result;
}
